README.md
# aws-core-test-utils v3.0.11
Utilities to assist with the unit testing of code using AWS

Modules:
- dynamodb-mocking.js - Utilities for generating mock DynamoDB.DocumentClient instances for unit testing
- kms-mocking.js - Utilities for generating mock AWS.KMS instances for unit testing
- samples.js - Utilities for generating samples of various AWS artifacts for testing

This module is exported as a [Node.js](https://nodejs.org) module.

## Installation

Using npm:
```bash
$ npm i aws-core-test-utils --save-dev
```

## Changes
See [CHANGES.md](CHANGES.md)

CHANGES.md
## Changes

### 3.0.11
- Changed `kms-mocking` module to enable dummy encryption & decryption of any arbitrary values

### 3.0.10
- Updated `tape` dev dependency to 4.9.1
- Updated `uuid` dev dependency to 3.2.1

### 3.0.9
- Removed broken & unwanted `core-functions` dependencies from `samples` module by including copies of needed 
  `core-functions` functions

### 3.0.8
- Added `.npmignore`
- Renamed `release_notes.md` to `CHANGES.md`
- Updated dependencies

### 3.0.7
- Minor fix to `README.md`

### 3.0.6
- Renamed dummy first exports (`exports._ = '_'; //IDE workaround`) of most modules to (`exports._$_ = '_$_';`) to avoid 
  potential future collisions with `lodash` & `underscore`

### 3.0.5
- Minor update to `node` engine version number in `package.json`

### 3.0.4
- Synchronized `samples.js` with version from `aws-core-utils` module

### 3.0.3
- Added dummy first exports (`exports._ = '_'; //IDE workaround`) to most modules as a temporary workaround for IDE issue

### 3.0.2
- Changed all modules' exports to modifications of the default `exports` object instead of replacing the default `module.exports` object

### 3.0.1
- Synchronized `samples.js` with versions from other modules

### 3.0.0
- Added `sampleKinesisMessageAndRecord` & `sampleDynamoDBMessageAndRecord` functions
- Removed `sampleKinesisMessage` function, which is no longer usable after removal of setting of `consumerState` 
  properties on messages

### 2.0.11
- Changes to `kms-mocking` module:
  - Changed mocked AWS.KMS `encrypt` & `decrypt` methods to also simulate the AWS.Request `send(callback)` & `promise()` 
    styles in addition to the existing immediate callback style
- Changes to `dynamodb-mocking` module:
  - Changed mocked DynamoDB.DocumentClient methods to also simulate the AWS.Request `send(callback)` & the immediate 
    callback styles in addition to the existing AWS.Request `promise()` style
- Removed `kms-utils` module, which should ONLY be in `aws-core-utils`

### 2.0.10
- Added new `kms-mocking` & `kms-utils` modules

### 2.0.9
- Moved test devDependencies to package.json & removed test/package.json
- Added `promises-extract` module with copy of `delay` function from `core-functions/promises` as at 2.0.16
- Removed `core-functions` dependency

### 2.0.8
- Added more unit tests for generate mock results case that generates simple item results

### 2.0.7
- Upgraded `uuid` test dependency to version 3.1.0

### 2.0.6
- Updated `core-functions` dependency to version 3.0.6
- Changes to `type-defs` module:
  - Added more properties to `TestObject`
  - Renamed `ValidateParams` type definition to `ValidateArgs` and changed it to accept variadic arguments
  - Renamed `MockResponse` type definition to `DynamoMockResponse` with new item `I` & key `K` templates
   - Changed its `result` property's type to `AnyDynamoResult`
   - Renamed its `validateParams` property to `validateArgs` & changed the property's type to `ValidateArgs`
  - Renamed `GenerateMockResponse` type definition to `GenerateDynamoMockResponse`
  - Added new `DynamoMockResponseSource` type definition
  - Added new `Params` type definition
  - Added new `AnyDynamoResult` type definition with item `I` & key `K` templates
  - Added new `AwsRequest` type definition with result `R` template
  - Renamed `DynamoDBDocClientMock` type definition to `DynamoDBDocClient` with new item `I` & key `K` templates
  - Improved definitions of `get`, `put`, `query` & `update` properties of `DynamoDBDocClient` type definition
  - Added new `batchGet`, `batchWrite`, `createSet`, `delete` & `scan` properties to `DynamoDBDocClient` type definition
  - Replaced all of the type definitions copied from the `module:aws-core-utils#type-defs` with their latest definitions
  
### 2.0.5
- Updated `core-functions` dependency to version 3.0.5

### 2.0.4
- Added support for also configuring a mock DynamoDB.DocumentClient with either an array of mock responses or a custom 
  generate mock response function instead of always just returning a single mock response

### 2.0.3
- Added simulation of DynamoDB.DocumentClient `update` method

### 2.0.2
- Added logic to avoid TypeErrors caused by attempts to access properties of undefined mock responses

### 2.0.1
- Fixed node version in `package.json`

### 2.0.0
- Initial commit to use with `core-functions` version 3.x

### 1.0.0
- Initial commit to use with `core-functions` version 2.x

test/samples.test.js
'use strict';

const test = require('tape');

const samples = require('../samples');

test('', t => {
  const shardId1 = 'shardId-111111111111';
  const seqNo1 = '49545115243490985018280067714973144582180062593244200961';
  const eventSourceARN = samples.sampleKinesisEventSourceArn('eu-west-1', 'TestStream');

  const [msg1, record1] = samples.sampleKinesisMessageAndRecord(shardId1, seqNo1, eventSourceARN, 'ID1', '1001', 'ABC',
    10, 1, 100, '10000000000000000000001', '2017-01-17T23:59:59.001Z');

  t.ok(msg1, `msg1 must exist`);
  t.ok(record1, `record1 must exist`);

  t.end();
});

test/kms-mocking.test.js
'use strict';

const test = require('tape');

// Test subject
const kmsMocking = require('../kms-mocking');
const mockKMS = kmsMocking.mockKMS;

// const util = require('util');

const accountId = 'XXXXXXXXXXXX';
const kmsKeyAlias = 'kkkkkkkkkkkkkk';
const region = 'eu-west-1';
const keyId = `arn:aws:kms:${region}:${accountId}:alias/${kmsKeyAlias}`;

const plaintext = 'Shhhhhhhhhhhhhhh';
const ciphertext = 'DUMMY_ciphertext';
const ciphertextBase64 = new Buffer(ciphertext, 'utf8').toString('base64');

// function show(o) {
//   return util.inspect(o);
// }

// ---------------------------------------------------------------------------------------------------------------------
// Immediate callback style
// ---------------------------------------------------------------------------------------------------------------------

test('mocking of KMS encrypt error using immediate callback style', t => {
  const encryptError = new Error('Encrypt kaboom');
  const kms = mockKMS(undefined, plaintext, encryptError, ciphertextBase64, 10);

  kms.encrypt({KeyId: keyId, Plaintext: plaintext}, (err, result) => {
    if (err) {
      t.pass(`encrypt must fail`);
      t.equal(err, encryptError, `encrypt error must be planned error`);
      t.end();
    } else {
      t.end(`encrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    }
  });
});

test('mocking of KMS decrypt error using immediate callback style', t => {
  const decryptError = new Error('Decrypt kaboom');
  const kms = mockKMS(decryptError, plaintext, undefined, ciphertextBase64, 10);

  kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')}, (err, result) => {
    if (err) {
      t.pass(`decrypt must fail`);
      t.equal(err, decryptError, `decrypt error must be planned error`);
      t.end();
    } else {
      t.end(`decrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    }
  });
});

test('mocking of KMS encrypt & decrypt using immediate callback style', t => {
  const kms = mockKMS(undefined, plaintext, undefined, ciphertextBase64, 10);

  kms.encrypt({KeyId: keyId, Plaintext: plaintext}, (err, result) => {
    if (err) {
      t.end(err);
    } else {
      const encrypted = result.CiphertextBlob && result.CiphertextBlob.toString('base64');
      t.ok(encrypted, `encrypted must be defined`);
      t.ok(encrypted.length > 0, `encrypted must be non-empty`);
      t.equal(encrypted, ciphertextBase64, `encrypted must be ciphertextBase64`);

      kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')}, (err, result) => {
        if (err) {
          t.end(err);
        } else {
          const decrypted = result.Plaintext && result.Plaintext.toString('utf8');
          // console.log(`##### decrypted = ${show(decrypted)}`);
          t.ok(decrypted, `decrypted must be defined`);
          t.ok(decrypted.length > 0, `decrypted must be non-empty`);
          t.equal(decrypted, plaintext, `decrypted must be plaintext`);
          t.end();
        }
      });
    }
  });
});

// ---------------------------------------------------------------------------------------------------------------------
// AWS.request.send(callback) style
// ---------------------------------------------------------------------------------------------------------------------

test('mocking of KMS encrypt error using AWS.request.send(callback) style', t => {
  const encryptError = new Error('Encrypt kaboom');
  const kms = mockKMS(undefined, plaintext, encryptError, ciphertextBase64, 10);

  const encryptRequest = kms.encrypt({KeyId: keyId, Plaintext: plaintext});
  encryptRequest.send((err, result) => {
    if (err) {
      t.pass(`encrypt must fail`);
      t.equal(err, encryptError, `encrypt error must be planned error`);
      t.end();
    } else {
      t.end(`encrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    }
  });
});

test('mocking of KMS decrypt error using AWS.request.send(callback) style', t => {
  const decryptError = new Error('Decrypt kaboom');
  const kms = mockKMS(decryptError, plaintext, undefined, ciphertextBase64, 10);

  const decryptRequest = kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')});
  decryptRequest.send((err, result) => {
    if (err) {
      t.pass(`decrypt must fail`);
      t.equal(err, decryptError, `decrypt error must be planned error`);
      t.end();
    } else {
      t.end(`decrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    }
  });
});

test('mocking of KMS encrypt & decrypt using AWS.request.send(callback) style', t => {
  const kms = mockKMS(undefined, plaintext, undefined, ciphertextBase64, 10);

  const encryptRequest = kms.encrypt({KeyId: keyId, Plaintext: plaintext});
  encryptRequest.send((err, result) => {
    if (err) {
      t.end(err);
    } else {
      const encrypted = result.CiphertextBlob && result.CiphertextBlob.toString('base64');
      t.ok(encrypted, `encrypted must be defined`);
      t.ok(encrypted.length > 0, `encrypted must be non-empty`);
      t.equal(encrypted, ciphertextBase64, `encrypted must be ciphertextBase64`);

      const decryptRequest = kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')});
      decryptRequest.send((err, result) => {
        if (err) {
          t.end(err);
        } else {
          const decrypted = result.Plaintext && result.Plaintext.toString('utf8');
          // console.log(`##### decrypted = ${show(decrypted)}`);
          t.ok(decrypted, `decrypted must be defined`);
          t.ok(decrypted.length > 0, `decrypted must be non-empty`);
          t.equal(decrypted, plaintext, `decrypted must be plaintext`);
          t.end();
        }
      });
    }
  });
});

// ---------------------------------------------------------------------------------------------------------------------
// AWS.request.promise() style
// ---------------------------------------------------------------------------------------------------------------------

test('mocking of KMS encrypt error using AWS.request.promise() style', t => {
  const encryptError = new Error('Encrypt kaboom');
  const kms = mockKMS(undefined, plaintext, encryptError, ciphertextBase64, 10);

  const encryptRequest = kms.encrypt({KeyId: keyId, Plaintext: plaintext});
  encryptRequest.promise().then(
    result => {
      t.end(`encrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    },
    err => {
      t.pass(`encrypt must fail`);
      t.equal(err, encryptError, `encrypt error must be planned error`);
      t.end();
    }
  );
});

test('mocking of KMS decrypt error using AWS.request.promise() style', t => {
  const decryptError = new Error('Decrypt kaboom');
  const kms = mockKMS(decryptError, plaintext, undefined, ciphertextBase64, 10);

  const decryptRequest = kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')});
  decryptRequest.promise().then(
    result => {
      t.end(`decrypt must NOT succeed with a result (${JSON.stringify(result)})`);
    },
    err => {
      t.pass(`decrypt must fail`);
      t.equal(err, decryptError, `decrypt error must be planned error`);
      t.end();
    }
  );
});

test('mocking of KMS encrypt & decrypt using AWS.request.promise() style', t => {
  const kms = mockKMS(undefined, plaintext, undefined, ciphertextBase64, 10);

  const encryptRequest = kms.encrypt({KeyId: keyId, Plaintext: plaintext});
  encryptRequest.promise().then(
    result => {
      const encrypted = result.CiphertextBlob && result.CiphertextBlob.toString('base64');
      // console.log(`##### encrypted = ${show(encrypted)}`);
      t.ok(encrypted, `encrypted must be defined`);
      t.ok(encrypted.length > 0, `encrypted must be non-empty`);
      t.equal(encrypted, ciphertextBase64, `encrypted must be ciphertextBase64`);

      const decryptRequest = kms.decrypt({CiphertextBlob: new Buffer(ciphertextBase64, 'base64')});
      decryptRequest.promise().then(
        result => {
          const decrypted = result.Plaintext && result.Plaintext.toString('utf8');
          // console.log(`##### decrypted = ${show(decrypted)}`);
          t.ok(decrypted, `decrypted must be defined`);
          t.ok(decrypted.length > 0, `decrypted must be non-empty`);
          t.equal(decrypted, plaintext, `decrypted must be plaintext`);
          t.end();
        },
        err => {
          t.end(err);
        }
      );
    },
    err => {
      t.end(err);
    }
  );
});

strings-extract.js
'use strict';

// ---------------------------------------------------------------------------------------------------------------------
// Excerpts from core-functions/strings
// ---------------------------------------------------------------------------------------------------------------------

exports.isString = isString;
exports.isNotBlank = isNotBlank;
exports.trim = trim;
exports.trimOrEmpty = trimOrEmpty;

/**
 * Returns true if the given value is a string; false otherwise.
 * @param {*} value - the value to check
 * @return {boolean} true if its a string; false otherwise
 */
function isString(value) {
  return typeof value === 'string' || value instanceof String;
}

/**
 * Returns true if the given string is NOT blank (i.e. NOT undefined, null, empty or contains only whitespace); false
 * otherwise.
 * @param {string|String} s - the string to check
 * @return {boolean} true if NOT blank; false otherwise
 */
function isNotBlank(s) {
  return s && (!s.trim || s.trim());
}

/**
 * Trims the given value if it is a string; otherwise returns a non-string value as is.
 * @param {*} value - the value to trim
 * @returns {string|*} the trimmed string or the original non-string value
 */
function trim(value) {
  return typeof value === 'string' || value instanceof String ? value.trim() : value;
}

/**
 * Trims the given value (if it's a string) or returns an empty string (if it's undefined or null); otherwise returns
 * the non-undefined, non-null, non-string value as is.
 * @param {*} value - the value to trim
 * @returns {string|*} the trimmed string; an empty string; or the original non-undefined, non-null, non-string value
 */
function trimOrEmpty(value) {
  return typeof value === 'string' || value instanceof String ? value.trim() :
    value === undefined || value === null ? '' : value;
}

samples.js
'use strict';

const Strings = require('./strings-extract');
const isNotBlank = Strings.isNotBlank;
const trim = Strings.trim;
const trimOrEmpty = Strings.trimOrEmpty;

const uuid = require('uuid');

const base64 = require('./base64-extract');

// Constants
const sampleAwsAccountId = "XXXXXXXXXXXX";
const sampleIdentityArn = 'identityarn';

const sampleFunctionName = "test-func";
const latestFunctionVersion = "$LATEST";

const sampleMessage = {
  key1: 'value1',
  key2: 'value2',
  key3: 'value3'
};

let nextSequenceNumber = 1;

/**
 * Utilities for generating samples of various AWS artifacts for testing.
 * @author Byron du Preez
 */
exports._$_ = '_$_'; //IDE workaround

exports.sampleAwsAccountId = sampleAwsAccountId;
exports.sampleIdentityArn = sampleIdentityArn;
exports.sampleFunctionName = sampleFunctionName;
exports.sampleMessage = sampleMessage;
exports.latestFunctionVersion = latestFunctionVersion;

  // General
exports.sampleNumberString = sampleNumberString;

  // For AWS contexts
exports.sampleInvokedFunctionArn = sampleInvokedFunctionArn;
exports.sampleAwsContext = sampleAwsContext;

  // For Kinesis events
exports.sampleStreamName = sampleStreamName;
exports.sampleKinesisEventSourceArn = sampleKinesisEventSourceArn;
exports.sampleKinesisEventSourceArnFromPrefixSuffix = sampleKinesisEventSourceArnFromPrefixSuffix;
exports.sampleBase64Data = sampleBase64Data;
exports.sampleKinesisPutRecordRequest = sampleKinesisPutRecordRequest;
exports.sampleKinesisRecord = sampleKinesisRecord;
exports.sampleKinesisEventWithSampleRecord = sampleKinesisEventWithSampleRecord;
exports.sampleKinesisEventWithRecord = sampleKinesisEventWithRecord;
exports.sampleKinesisEventWithRecords = sampleKinesisEventWithRecords;

exports.sampleMsg = sampleMsg;
exports.sampleKinesisRecord2 = sampleKinesisRecord2;
exports.sampleKinesisMessageAndRecord = sampleKinesisMessageAndRecord;

exports.awsKinesisStreamsSampleEvent = awsKinesisStreamsSampleEvent;

  // For DynamoDB stream events
exports.sampleTableName = sampleTableName;
exports.sampleDynamoDBEventSourceArn = sampleDynamoDBEventSourceArn;
exports.sampleDynamoDBEventSourceArnFromPrefixSuffix = sampleDynamoDBEventSourceArnFromPrefixSuffix;

exports.sampleDynamoDBRecord = sampleDynamoDBRecord;
exports.sampleDynamoDBMessageAndRecord = sampleDynamoDBMessageAndRecord;

exports.sampleDynamoDBEventWithRecords = sampleDynamoDBEventWithRecords;

exports.awsDynamoDBUpdateSampleEvent = awsDynamoDBUpdateSampleEvent;


// ---------------------------------------------------------------------------------------------------------------------
// Start of code borrowed from `aws-core-utils/arns`
// ---------------------------------------------------------------------------------------------------------------------
const REGION_INDEX = 3;
/**
 * Extracts the component at the given index of the given ARN.
 * @param {string} arn the ARN from which to extract the component
 * @param {number} index the index of the ARN component to retrieve
 * @returns {string} the component (if extracted) or empty string (if not)
 */
function getArnComponent(arn, index) {
  const components = isNotBlank(arn) ? trim(arn).split(':') : [];
  return components.length > index ? trimOrEmpty(components[index]) : '';
}
// ---------------------------------------------------------------------------------------------------------------------
// End of code borrowed from `aws-core-utils/arns`
// ---------------------------------------------------------------------------------------------------------------------

function sampleNumberString(digits) {
  let number = "";
  for (let i = 0; i < digits; ++i) {
    number += Math.floor((Math.random() * 10)); // 0 to 9
  }
  return number;
}

function sampleStreamName(streamNamePrefix, streamNameSuffix) {
  const prefix = isNotBlank(streamNamePrefix) ? streamNamePrefix : 'TestKinesisStream';
  const suffix = isNotBlank(streamNameSuffix) ? streamNameSuffix : '';
  return `${prefix}${suffix}`;
}

function sampleTableName(tableNamePrefix, tableNameSuffix) {
  const prefix = isNotBlank(tableNamePrefix) ? tableNamePrefix : 'TestDynamoDBTable';
  const suffix = isNotBlank(tableNameSuffix) ? tableNameSuffix : '';
  return `${prefix}${suffix}`;
}

function sampleInvokedFunctionArn(invokedFunctionArnRegion, functionName, functionAlias) {
  const region = isNotBlank(invokedFunctionArnRegion) ? invokedFunctionArnRegion : 'IF_ARN_REGION';
  const funcName = isNotBlank(functionName) ? functionName : sampleFunctionName;
  const aliasSuffix = isNotBlank(functionAlias) ? `:${functionAlias}` : '';
  return `arn:aws:lambda:${region}:${sampleAwsAccountId}:function:${funcName}${aliasSuffix}`
}

function sampleKinesisEventSourceArn(eventSourceArnRegion, streamName) {
  const region = isNotBlank(eventSourceArnRegion) ? eventSourceArnRegion : 'EF_ARN_REGION';
  const streamName1 = isNotBlank(streamName) ? streamName : sampleStreamName();
  return `arn:aws:kinesis:${region}:${sampleAwsAccountId}:stream/${streamName1}`;
}

function sampleDynamoDBEventSourceArn(eventSourceArnRegion, tableName, timestamp) {
  const region = isNotBlank(eventSourceArnRegion) ? eventSourceArnRegion : 'EF_ARN_REGION';
  const tableName1 = isNotBlank(tableName) ? tableName : sampleTableName();
  const timestamp0 = isNotBlank(timestamp) ? timestamp : new Date().toISOString();
  const timestamp1 = timestamp0.endsWith('Z') ? timestamp0.substring(0, timestamp0.length - 1) : timestamp0;
  //arn:aws:dynamodb:us-east-1:111111111111:table/test/stream/2020-10-10T08:18:22.385
  return `arn:aws:dynamodb:${region}:${sampleAwsAccountId}:table/${tableName1}/stream/${timestamp1}`;
}

function sampleKinesisEventSourceArnFromPrefixSuffix(eventSourceArnRegion, streamNamePrefix, streamNameSuffix) {
  const streamName = sampleStreamName(streamNamePrefix, streamNameSuffix);
  return sampleKinesisEventSourceArn(eventSourceArnRegion, streamName);
}

function sampleDynamoDBEventSourceArnFromPrefixSuffix(eventSourceArnRegion, tableNamePrefix, tableNameSuffix, timestamp) {
  const tableName = sampleTableName(tableNamePrefix, tableNameSuffix);
  return sampleDynamoDBEventSourceArn(eventSourceArnRegion, tableName, timestamp);
}

function sampleAwsContext(functionName, functionVersion, invokedFunctionArn, maxTimeInMillis) {
  const uuid1 = uuid.v4();
  const startTime = Date.now();
  const maximumTimeInMillis = maxTimeInMillis ? maxTimeInMillis : 1000;

  const dashRegex = /-/g;
  dashRegex.lastIndex = 0; //NB: MUST RESET lastIndex for global regular expressions (i.e. /.../g )
  //noinspection JSUnusedLocalSymbols
  const logStreamName = `2016/10/14/[$LATEST]${uuid1.replace(dashRegex, "")}`;
  dashRegex.lastIndex = 0; //NB: MUST RESET lastIndex for global regular expressions (i.e. /.../g )

  return {
    callbackWaitsForEmptyEventLoop: true,
    logGroupName: `/aws/lambda/${functionName}`,
    logStreamName: logStreamName,
    functionName: functionName,
    memoryLimitInMB: 128,
    functionVersion: functionVersion,
    invokeid: uuid1,
    awsRequestId: uuid1,
    invokedFunctionArn: invokedFunctionArn,
    getRemainingTimeInMillis() {
      return maximumTimeInMillis - (Date.now() - startTime);
    }
  };
}

function sampleBase64Data(obj) {
  return new Buffer(JSON.stringify(obj), 'base64');
}

function sampleKinesisPutRecordRequest(streamName, partitionKey, data, explicitHashKey, sequenceNumberForOrdering) {
  // Data "SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=" is a harmless message ('Hello, this is a test 123.')
  // var c = new Buffer('SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=', 'base64').toString('utf-8');
  // var d = new Buffer('Hello, this is a test 123.', 'utf-8').toString('base64');
  const kinesisPartitionKey = isNotBlank(partitionKey) ? partitionKey : uuid.v4();
  const kinesisData = data !== undefined ?
    typeof data === 'object' ? JSON.stringify(data) : data : "SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=";
  const putRecordRequest = {
    StreamName: streamName,
    PartitionKey: kinesisPartitionKey,
    Data: kinesisData
  };
  if (explicitHashKey) {
    putRecordRequest.ExplicitHashKey = explicitHashKey;
  }
  if (sequenceNumberForOrdering) {
    putRecordRequest.SequenceNumberForOrdering = sequenceNumberForOrdering;
  }
  return putRecordRequest;
}

function sampleKinesisRecord(shardId, sequenceNumber, partitionKey, data, eventSourceArn, eventAwsRegion) {
  // Data "SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=" is a harmless message ('Hello, this is a test 123.')
  // var c = new Buffer('SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=', 'base64').toString('utf-8');
  // var d = new Buffer('Hello, this is a test 123.', 'utf-8').toString('base64');
  // const seqNo = sampleNumberString(56);
  const kinesisPartitionKey = isNotBlank(partitionKey) ? partitionKey : uuid.v4();
  const kinesisData = data !== undefined ?
    typeof data === 'object' ? base64.toBase64(data) : data : "SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=";
  sequenceNumber = sequenceNumber ? sequenceNumber : nextSequenceNumber++; //sampleNumberString(56);
  const eventSourceRegion = eventSourceArn ? getArnComponent(eventSourceArn, REGION_INDEX) : '';
  const awsRegion = eventAwsRegion ? eventAwsRegion : eventSourceRegion ? eventSourceRegion : 'us-west-2';
  shardId = shardId ? shardId : 'shardId-000000000000';
  const event = {
    eventID: `${shardId}:${sequenceNumber}`,
    eventVersion: "1.0",
    kinesis: {
      partitionKey: kinesisPartitionKey,
      data: kinesisData,
      kinesisSchemaVersion: "1.0",
      sequenceNumber: `${sequenceNumber}`
    },
    invokeIdentityArn: sampleIdentityArn,
    eventName: "aws:kinesis:record",
    //eventSourceARN: eventSourceArn,
    eventSource: "aws:kinesis",
    awsRegion: awsRegion
  };
  if (eventSourceArn !== undefined) {
    event.eventSourceARN = eventSourceArn;
  }
  return event;
}

function sampleKinesisEventWithSampleRecord(shardId, sequenceNumber, partitionKey, data, eventSourceArn, eventAwsRegion) {
  return sampleKinesisEventWithRecord(sampleKinesisRecord(shardId, sequenceNumber, partitionKey, data, eventSourceArn, eventAwsRegion))
}

function sampleKinesisEventWithRecord(kinesisRecord) {
  return {
    Records: [
      kinesisRecord
    ]
  };
}

function sampleKinesisEventWithRecords(kinesisRecords) {
  return {
    Records: kinesisRecords
  };
}

function sampleMsg(id1, id2, k1, k2, n1, n2, n3, n4, n5) {
  const msg = {};
  if (id1) msg.id1 = id1;
  if (id2) msg.id2 = id2;

  if (k1) msg.k1 = k1;
  if (k2) msg.k2 = k2;

  if (n1) msg.n1 = n1;
  if (n2) msg.n2 = n2;
  if (n3) msg.n3 = n3;
  if (n4) msg.n4 = n4;
  if (n5) msg.n5 = n5;
  return msg;
}

/**
 * @returns {[Message, Record]}
 */
function sampleKinesisMessageAndRecord(shardId, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5) {
  const msg = sampleMsg(id1, id2, k1, k2, n1, n2, n3, n4, n5);

  const record = sampleKinesisRecord(shardId, eventSeqNo, undefined, msg, eventSourceARN, undefined);

  return [msg, record];
}

function sampleKinesisRecord2(shardId, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5) {
  const msg = sampleMsg(id1, id2, k1, k2, n1, n2, n3, n4, n5);

  return sampleKinesisRecord(shardId, eventSeqNo, undefined, msg, eventSourceARN, undefined);
}

function awsKinesisStreamsSampleEvent(identityArn, eventSourceArn) {
  return {
    "Records": [
      {
        "eventID": "shardId-000000000000:49545115243490985018280067714973144582180062593244200961",
        "eventVersion": "1.0",
        "kinesis": {
          "partitionKey": "partitionKey-3",
          "data": "SGVsbG8sIHRoaXMgaXMgYSB0ZXN0IDEyMy4=",
          "kinesisSchemaVersion": "1.0",
          "sequenceNumber": "49545115243490985018280067714973144582180062593244200961"
        },
        "invokeIdentityArn": identityArn,
        "eventName": "aws:kinesis:record",
        "eventSourceARN": eventSourceArn,
        "eventSource": "aws:kinesis",
        "awsRegion": "us-east-1"
      }
    ]
  };
}

function sampleDynamoDBRecord(eventID, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5) {
  const record = {
    "eventID": eventID,
    "eventVersion": "1.0",
    "dynamodb": {
      "StreamViewType": "NEW_AND_OLD_IMAGES",
      "SequenceNumber": eventSeqNo,
      "SizeBytes": 26
    },
    "awsRegion": "us-west-2",
    "eventName": "INSERT",
    "eventSourceARN": eventSourceARN,
    "eventSource": "aws:dynamodb"
  };
  const dynamodb = record.dynamodb;
  if (k1 || k2) {
    dynamodb.Keys = {};
  }
  if (k1 || k2 || id1 || id2 || n1 || n2 || n3 || n4 || n5) {
    dynamodb.NewImage = {};
    dynamodb.OldImage = {};
  }
  const keys = dynamodb.Keys;
  const newImage = dynamodb.NewImage;
  const oldImage = dynamodb.OldImage;
  if (k1) {
    keys.k1 = {"S": `${k1}`};
    newImage.k1 = {"S": `${k1}`};
    oldImage.k1 = {"S": `${k1}`};
  }
  if (k2) {
    keys.k2 = {"N": `${k2}`};
    newImage.k2 = {"N": `${k2}`};
    oldImage.k2 = {"N": `${k2}`};
  }
  if (id1) {
    newImage.id1 = {"S": `${id1}`};
    oldImage.id1 = {"S": `${id1}`};
  }
  if (id2) {
    newImage.id2 = {"S": `${id2}`};
    oldImage.id2 = {"S": `${id2}`};
  }
  if (n1) {
    newImage.n1 = {"N": `${n1}`};
    oldImage.n1 = {"N": `${n1}`};
  }
  if (n2) {
    newImage.n2 = {"N": `${n2}`};
    oldImage.n2 = {"N": `${n2}`};
  }
  if (n3) {
    newImage.n3 = {"N": `${n3}`};
    oldImage.n3 = {"N": `${Number(n3) - 1}`};
  }
  if (n4) {
    newImage.n4 = {"S": `${n4}`};
    oldImage.n4 = {"S": `${n4}`};
  }
  if (n5) {
    newImage.n5 = {"S": `${n5}`};
    oldImage.n5 = {"S": `${n5}`};
  }
  return record;
}

/**
 * @returns {[Message, Record]}
 */
function sampleDynamoDBMessageAndRecord(eventID, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5) { // , skipSimplify
  const record = sampleDynamoDBRecord(eventID, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5);

  // const msg = copy(record, deep);
  // Create a "copy" of record without `core-functions/copying#copy`
  const msg = sampleDynamoDBRecord(eventID, eventSeqNo, eventSourceARN, id1, id2, k1, k2, n1, n2, n3, n4, n5);

  // if (!skipSimplify) {
  //   dynamoDBUtils.simplifyKeysNewImageAndOldImage(msg.dynamodb);
  // }
  return [msg, record];
}

function sampleDynamoDBEventWithRecords(dynamoDBRecords) {
  return {
    Records: dynamoDBRecords
  };
}

function awsDynamoDBUpdateSampleEvent(eventSourceArn) {
  return {
    "Records": [
      {
        "eventID": "1",
        "eventVersion": "1.0",
        "dynamodb": {
          "Keys": {
            "Id": {
              "N": "101"
            }
          },
          "NewImage": {
            "Message": {
              "S": "New item!"
            },
            "Id": {
              "N": "101"
            }
          },
          "StreamViewType": "NEW_AND_OLD_IMAGES",
          "SequenceNumber": "111",
          "SizeBytes": 26
        },
        "awsRegion": "us-west-2",
        "eventName": "INSERT",
        "eventSourceARN": eventSourceArn,
        "eventSource": "aws:dynamodb"
      },
      {
        "eventID": "2",
        "eventVersion": "1.0",
        "dynamodb": {
          "OldImage": {
            "Message": {
              "S": "New item!"
            },
            "Id": {
              "N": "101"
            }
          },
          "SequenceNumber": "222",
          "Keys": {
            "Id": {
              "N": "101"
            }
          },
          "SizeBytes": 59,
          "NewImage": {
            "Message": {
              "S": "This item has changed"
            },
            "Id": {
              "N": "101"
            }
          },
          "StreamViewType": "NEW_AND_OLD_IMAGES"
        },
        "awsRegion": "us-west-2",
        "eventName": "MODIFY",
        "eventSourceARN": eventSourceArn,
        "eventSource": "aws:dynamodb"
      },
      {
        "eventID": "3",
        "eventVersion": "1.0",
        "dynamodb": {
          "Keys": {
            "Id": {
              "N": "101"
            }
          },
          "SizeBytes": 38,
          "SequenceNumber": "333",
          "OldImage": {
            "Message": {
              "S": "This item has changed"
            },
            "Id": {
              "N": "101"
            }
          },
          "StreamViewType": "NEW_AND_OLD_IMAGES"
        },
        "awsRegion": "us-west-2",
        "eventName": "REMOVE",
        "eventSourceARN": eventSourceArn,
        "eventSource": "aws:dynamodb"
      }
    ]
  };
}